### PR TITLE
Example monte-carlo-integration

### PR DESCRIPTION
**Pull request by jkrude**
This is an _alpaka_ implementation of the **monte-carlo-integration** method.

The example targets the topics:
- (pseudo) random-number-generation
- atomic functions

The code is based on a cpu-prototype by @sbastrakov.

### CODE REVIEW

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+
+#include <iostream>
+#include <typeinfo>
+
+
+//#############################################################################
+//! This functor defines the function for which the integral is to computed.
```
Comment (BenjaminW3) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]:
```suggestion
//! This functor defines the function for which the integral is to be computed.
```

**Review by psychocoderHPC — APPROVED** [review]

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
+ * IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <alpaka/alpaka.hpp>
+#include <alpaka/example/ExampleDefaultAcc.hpp>
+
+#include <iostream>
+#include <typeinfo>
```
Comment (psychocoderHPC) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: please include `<cstdlib>` for `EXIT_SUCCESS` and `EXIT_FAILURE`

**Review by psychocoderHPC — CHANGES_REQUESTED** [review]

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
+/* Copyright 2019 Benjamin Worpitz, Sergei Bastrakov, Jakob Krude
```
Comment (sbastrakov) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: I think the year should be 2020?

Or what is alpaka's policy on that. The file is new, but based on other examples from 2019.

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    //! \param x The argument.
+    template<typename TAcc>
+    ALPAKA_FN_ACC
+    auto operator ()(
+        TAcc const & acc,
+        float const x) -> float
+    {
+        return alpaka::math::sqrt(
+            acc,
+            (1 - x * x));
```
Comment (sbastrakov) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: 1.0f is a little cleaner than 1.

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            alpaka::Threads>(acc);
+
+        auto const linearizedGlobalThreadIdx = alpaka::idx::mapIdx<1u>(
+            globalThreadIdx,
+            globalThreadExtent);
+        // Setup generator and distribution.
+        auto generator = alpaka::rand::generator::createDefault(
+            acc,
+            linearizedGlobalThreadIdx[0],
+            0);
```
Comment (sbastrakov) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: Perhaps it is easier to make `linearizedGlobalThreadIdx ` be a single number, by adding `[0]` at line 79 before `;`.

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        // For simplicity the interval is fixed at [0,1].
+        auto dist(alpaka::rand::distribution::createUniformReal<float>(acc));
+
+        uint32_t localCount = 0;
+        for(size_t i = globalThreadIdx[0];
```
Comment (sbastrakov) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: I think this is a mistake and should be `linearizedGlobalThreadIdx[0]` (or without `[0]` with my earlier suggestion) instead.

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    alpaka::mem::view::copy(
+        queue,
+        bufHost,
+        bufAcc,
+        extent);
+    alpaka::wait::wait(queue);
+
+    // Check the result.
+    uint32_t globalCount = *ptrBufHost;
+    std::cout << "global count: " << globalCount << "\n";
```
Comment (sbastrakov) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: I think this output could be removed. As for a user the latter output of exact and actual result is informative, but this one is not necessarily.

**Review by sbastrakov — CHANGES_REQUESTED** [review]

**Review by sbastrakov — DISMISSED** [review]

File: example/monteCarloIntegration/src/monteCarloIntegration.cpp [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+        auto const linearizedGlobalThreadIdx = alpaka::idx::mapIdx<1u>(
+            globalThreadIdx,
+            globalThreadExtent)[0];
+        // Setup generator and distribution.
+        auto generator = alpaka::rand::generator::createDefault(
+            acc,
+            linearizedGlobalThreadIdx,
+            0); // No specific subsequence start.
+        // For simplicity the interval is fixed at [0,1].
```
Comment (BenjaminW3) [inline example/monteCarloIntegration/src/monteCarloIntegration.cpp]: Documentation of `createUniformReal` says `[0.0, 1.0)`.

Comment (sbastrakov) [reply example/monteCarloIntegration/src/monteCarloIntegration.cpp]: Yes. So for Monte-Carlo does not matter, but better be precise at the comment.

**Review by BenjaminW3 — APPROVED** [review]

**Pull request merged**